### DISCUSSION TITLE
Adding a DataLoader without the Annotation?

### DISCUSSION THREAD

**Discussion by evanmalmud** [Netflix/dgs-framework#1685] [Q&A] [answered]
Hello trying to setup a dataloader with slightly different options than default so I made my own bean. Mainly trying to utilize the cacheMap feature. This works to a point but now
```
  @Bean
  public DataLoader<String, Map<String, Object>> blockDataLoader() {
    DataLoaderOptions options = new DataLoaderOptions();
    options.setBatchingEnabled(true);
    options.setCacheMap((CacheMap<?, ?>) springCacheValueCache);
    MappedBatchLoaderWithContext loader =
        new BlockDataLoader(storefrontBlockRepository, blockDataPostProcessors, scheduledPreviewService);
    return DataLoader.newMappedDataLoader(loader, options);
  }
```

Im having trouble making sure this dataloader is added to the the DgsDataLoaderRegistry, ultimately i cant use the "DgsDataLoader" annotation because that tries to make a second bean, and I dont see a way to add the "name" or be picked up here by the DgsAutoconfiguration. 

Would love some insight on the best path forward. Essentially all I want to do is set options for a DgsDataLoader.
Im stuck on Dgs 5 btw. (Cant upgrade spring yet)

**Comment by srinivasankavitha** [comment]
Hi @evanmalmud - we don't currently have a way to customize with your own dataloader unfortunately. This is the case with both 5.x and later releases. We will need to implement support for this feature by enhancing the existing @DgsDataloader annotation to allow the cache map to be configurable. We can consider adding that in a future release, but due to other priorities we won't be able to deliver this at the moment. Feel free to open a PR and we'd be happy to review it.

**Reply by evanmalmud** [reply]
Is this the viable path forward? 
https://github.com/Netflix/dgs-framework/pull/202
Happy to update this. Im going to try shading for now.

Also seems like you wouldn't be moving this to the 2.7 branch anyway? based on other conversations i've seen?
This would mean we have to make our own version?

**Reply by srinivasankavitha** [reply]
I think the easiest/low risk path forward is to just expose this as a configurable setting in the data loader annotation. Let me see if I can open a quick PR in the next few days for that.

**Reply by evanmalmud** [reply]
That would be very helpful. Thanks so much.

Might as well include the other options if you are in there for other too!

**Reply by srinivasankavitha** [reply]
Actually, it appears we already have some hooks in place. Could you please try defining this with your custom options.
```class CustomDataLoaderOptionsProvider : DgsDataLoaderOptionsProvider {
    override fun getOptions(dataLoaderName: String, annotation: DgsDataLoader): DataLoaderOptions {
        val options = DataLoaderOptions()
            .setBatchingEnabled(annotation.batching)
            .setCachingEnabled(annotation.caching)
        if (annotation.maxBatchSize > 0) {
            options.setMaxBatchSize(annotation.maxBatchSize)
        }
        return options
    }
}


You can then create a bean for the DataLoaderOptions like so:
```
    @Bean
    @ConditionalOnMissingBean
    open fun dgsDataLoaderOptionsProvider(): DgsDataLoaderOptionsProvider {
        return CustomDataLoaderOptionsProvider()
    }

```

```

**Reply by evanmalmud** [reply]
I'm looking and it seems this change was never brought to the 2.7 branch?

**Reply by srinivasankavitha** [reply]
Not it's only available on the main 3.x branch. We are not introducing new features in the 2.7 branch since it is in maintenance mode.

**Reply by evanmalmud** [reply]
Alright, so is there anyway to do this on 2.7? Honestly even just this minor feature would be greatly appreciated. 
We are working to go to 3.0, but i dont think we can reasonably do it for a few months.

**Reply by srinivasankavitha** [reply]
Ok, since it was one commit, I ported it over to the 2.7 branch. The release `5.6.0` is available with the same change. Please switch over to the 3.x compatible releases when you are able to so you can get the latest fixes and improvements from there going forward.

**Reply by evanmalmud** [reply]
Thanks so much! This will greatly help us!

**Reply by evanmalmud** [reply]
Hey just checking in here. trying to use this new feature and I can't see to get the autoconfigruation to work and looking at the code im unsure how its supposed to.

So ive made my own bean of which should be overriding the default,
but looking here where the provider is setup its not using the bean?
https://github.com/Netflix/dgs-framework/blob/3afa77761d774a1c947c6aad56d5aca13dd63030/graphql-dgs-spring-boot-oss-autoconfigure/src/main/kotlin/com/netflix/graphql/dgs/autoconfig/DgsAutoConfiguration.kt#L157
^ Here its not taking in an Autowire of the the OptionsProvider

and the default for it doesnt seem to be grabbing a bean either.
https://github.com/Netflix/dgs-framework/blob/3afa77761d774a1c947c6aad56d5aca13dd63030/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsDataLoaderProvider.kt#L49

I'm not super familiar with kotlin so I could be totally off base here. But im just having trouble getting this working.


Just to spell out what I've done
Created an impl of DgsDataLoaderOptionsProvider and set some value and caching!
```
public class DataLoaderConfig implements DgsDataLoaderOptionsProvider {

  @NotNull
  @Override
  public DataLoaderOptions getOptions(@NotNull String s, @NotNull DgsDataLoader dgsDataLoader) {
    return DataLoaderOptions.newOptions()
        .setBatchingEnabled(true)
        .setCachingEnabled(true)
        .setValueCache(new SpringValueMap<>())
        .setCacheMap(new SpringCacheMap<>());

  }
  ```
And then just created a bean of the above

```
  @Bean
  public DgsDataLoaderOptionsProvider dgsDataLoaderOptionsProvider() {
    return new DataLoaderConfig();
  }
  ```
  
  It didn't seem like there was anything more to do, but im missing something.

**Reply by srinivasankavitha** [reply]
Could you try `5.6.1` ? I put in a fix for the issue you pointed out.

**Reply by evanmalmud** [reply]
Works great thank you!

**Reply by evanmalmud** [reply]
Updating the answer here for others.

**Comment by evanmalmud** [comment, marked as answer]
Thanks to @srinivasankavitha you can now override options for dataloaders by making a DgsDataLoaderOptionsProvider bean

Create an implementation of DgsDataLoaderOptionsProvider 
```
public class DataLoaderConfig implements DgsDataLoaderOptionsProvider {

  @NotNull
  @Override
  public DataLoaderOptions getOptions(@NotNull String s, @NotNull DgsDataLoader dgsDataLoader) {
    return DataLoaderOptions.newOptions()
        //Whatever options you want to set
        ;
  }
```

Create a bean of this class that will be automatically picked up by DGS
```
  @Bean
  public DgsDataLoaderOptionsProvider dgsDataLoaderOptionsProvider() {
    return new DataLoaderConfig();
  }
```

**Comment by deep18-appdirect** [comment]
Hi @evanmalmud 
I followed the same thing but I think `dgsDataLoaderProvider` still taking `DefaultDataLoaderOptionsProvider` as I have added log in `DataLoaderConfig` but it's never logged.
can you please let me know how exactly you are handling this.

**Reply by deep18-appdirect** [reply]
Working now...
it was not working in the version 7.1.0, need to update the version.
